Message edit/delete overhaul

liwilliam2021 opened the pull request:
Four big changes (note this feature is still disabled until checkpoints are overhauled)
1. Made the edit interface a variation of the Chat Area 
2. Use our own modal instead of native vscode modal for delete/edit warnings
3. Removed the option to "delete just this message" and replaced with a warning
~~4. Added an option to stop showing edit/delete warnings~~
<img width="369" alt="Screenshot 2025-07-09 at 3 25 11 PM" src="https://github.com/user-attachments/assets/55d3db43-b5a1-45f4-9039-4509386ec21c" />

Partially addresses #4703 

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Overhauls message edit/delete functionality with custom modals, removes individual delete option, and updates edit interface to resemble chat area.
> 
>   - **Behavior**:
>     - Overhauls message edit/delete functionality with a custom modal for confirmations in `webviewMessageHandler.ts`.
>     - Removes "delete just this message" option, replacing it with a warning.
>     - Adds option to stop showing edit/delete warnings.
>   - **UI Changes**:
>     - Updates edit interface to resemble chat area in `App.tsx`.
>     - Introduces `DeleteMessageDialog` and `EditMessageDialog` components in `App.tsx`.
>   - **Testing**:
>     - Updates tests in `ClineProvider.spec.ts` to reflect new delete/edit behavior.
>     - Adds tests for edit mode integration in `ChatTextArea.spec.tsx` and `EditModeControls.spec.tsx`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for cf53be74ab5ccba81429b300b8464886cd1f96cb. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

daniel-lxs commented:
Hey @liwilliam2021, I gave the implementation a try and noticed a couple of issues. It looks like editing a message and attaching a new image doesn’t work as expected right now.

Also, I think we could simplify the UI a bit by removing the codebase indexing status indicator, and adjust the styles so the edit message text area feels more integrated. Right now it looks a bit like a box inside a box.

Let me know if anything's unclear or if you’d like to talk through any of this.

liwilliam2021 commented:
> Hey @liwilliam2021, I gave the implementation a try and noticed a couple of issues. It looks like editing a message and attaching a new image doesn’t work as expected right now.
> 
> Also, I think we could simplify the UI a bit by removing the codebase indexing status indicator, and adjust the styles so the edit message text area feels more integrated. Right now it looks a bit like a box inside a box.
> 
> Let me know if anything's unclear or if you’d like to talk through any of this.

hmmm weird-- they seemed to work locally on my end? thanks for the feedback will check it out!

liwilliam2021 commented:
Thanks @daniel-lxs -- those are great catches! 

I also removed the skip warning dialog option bc it will be replaced in the next PR anyways.
<img width="441" height="193" alt="Screenshot 2025-07-10 at 8 26 25 PM" src="https://github.com/user-attachments/assets/e9911a8a-3d51-42d5-b19d-436fc4ab0f91" />
<img width="437" height="130" alt="Screenshot 2025-07-10 at 7 32 41 PM" src="https://github.com/user-attachments/assets/be5d6b0d-8518-4b07-9651-019b7fa61d48" />

liwilliam2021 commented:
> Hey Will, thank you for addressing the suggestions.
> 
> I noticed that the model doesn't seem to be aware of the images being sent:
> 
> <img alt="image" width="396" height="280" src="https://private-user-images.githubusercontent.com/57051444/466229164-76d12dff-2b78-4dd1-9f05-71a6fbb01927.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI1OTUwMzgsIm5iZiI6MTc1MjU5NDczOCwicGF0aCI6Ii81NzA1MTQ0NC80NjYyMjkxNjQtNzZkMTJkZmYtMmI3OC00ZGQxLTlmMDUtNzFhNmZiYjAxOTI3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE1VDE1NTIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhNWRjOWJkOTQ5MTNiNTg3MTdjMDNkODQzZTU5ZTZmZDVjZTBkMmRkMWI2Zjc1YjAyNmZiZDAzYjU5MGNhOWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.JyEkdxgsU2-8UBz7obLhsu0B41PvxdPXhDR-UigOdj4">
> Maybe I'm doing something wrong here?
> 
> Also, I think the padding around the text area could be a bit smaller:
> 
> <img alt="image" width="404" height="611" src="https://private-user-images.githubusercontent.com/57051444/466229380-bbfe3384-4459-4191-94b1-7242b8cd9628.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI1OTUwMzgsIm5iZiI6MTc1MjU5NDczOCwicGF0aCI6Ii81NzA1MTQ0NC80NjYyMjkzODAtYmJmZTMzODQtNDQ1OS00MTkxLTk0YjEtNzI0MmI4Y2Q5NjI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzE1VDE1NTIxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVmNTQzN2QzOWU2NzlhNjQ0ZTA3Yjc5MDY3MDZjZmZkY2Y2YjM3NzE3YmE2YmY0NWE4NWYyODZkMzZkYjQ0MmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.B3QIeYC6zz4lds1WhsQCO8Sr6rr4mcnykFm_h6Qlvq0">

hmmm-- I can fix the padding but the image issue doesn't seem to appear on my end

https://www.loom.com/share/156cd936f3bf432887408f53e717789c

daniel-lxs commented on webview-ui/src/App.tsx:
```diff
@@ -66,6 +67,26 @@ const App = () => {
+		messageTs: 0,
+		text: "",
+		images: [],
+	})
```
Consider defining explicit TypeScript interfaces for the dialog state objects to improve type safety:

```typescript
interface DeleteMessageDialogState {
  isOpen: boolean;
  messageTs: number;
}

interface EditMessageDialogState {
  isOpen: boolean;
  messageTs: number;
  text: string;
  images?: string[];
}
```

This would make the code more maintainable and catch potential type errors at compile time.

daniel-lxs commented on webview-ui/src/App.tsx:
```diff
@@ -207,6 +241,30 @@ const App = () => {
+					})
+					setEditMessageDialogState((prev) => ({ ...prev, isOpen: false }))
+				}}
+			/>
```
Consider memoizing the dialog components to prevent unnecessary re-renders:

```typescript
const MemoizedDeleteMessageDialog = React.memo(DeleteMessageDialog);
const MemoizedEditMessageDialog = React.memo(EditMessageDialog);
```

This could improve performance, especially when the parent component re-renders frequently.

daniel-lxs commented on webview-ui/src/components/chat/ChatTextArea.tsx:
```diff
@@ -779,6 +784,449 @@ const ChatTextArea = forwardRef<HTMLTextAreaElement, ChatTextAreaProps>(
+					</StandardTooltip>
+				</div>
+			</div>
+		)
```
The edit mode rendering logic is quite complex with multiple helper functions. Consider extracting the edit mode UI into a separate component for better maintainability:

```typescript
const EditModeControls: React.FC<EditModeControlsProps> = ({ ... }) => {
  // Edit mode specific UI logic
};
```

This would simplify the main component and make the code more modular.

daniel-lxs commented on webview-ui/src/i18n/locales/en/common.json:
```diff
@@ -51,5 +51,13 @@
+		"edit_message": "Edit Message",
+		"edit_warning": "Editing this message will delete all subsequent messages in the conversation. Do you want to proceed?",
+		"proceed": "Proceed",
+		"dont_show_again": "Don't show this again"
```
I noticed a few keys in the code that use underscores in their names. It’s probably a good idea to standardize or update these to match the naming conventions used elsewhere in the the translations files. This could help avoid confusion and keep things consistent.

liwilliam2021 replied:
good catch!

daniel-lxs reviewed (approved):
I think this is working correctly. I'll leave a few minor suggestions, but I think it's close to being done.

KJ7LNW commented:
am I missing something?  how do I edit? 

I see the trash can icon but I do not see the edit icon:

<img width="586" height="203" alt="image" src="https://github.com/user-attachments/assets/f7553bf7-9f05-4ac4-a87a-7d3769aa8b8a" />
